File: exercises/ex2_occasional/occasional.cpp
#include <chrono>
#include <cstdlib>
#include <ctime>
#include <iostream>
#include <thread>

void decoyFunction() {
  // This function does nothing but serves as a decoy
  return;
}

int main() {
  // Seed the random number generator
  std::srand(std::time(0));

  // Generate a random number between 0 and 99
  int randomNumber = std::rand() % 100;

  std::cout << "Generated number: " << randomNumber << std::endl;

  decoyFunction(); // Call the decoy function

  // Sleep for 1 second
  std::this_thread::sleep_for(std::chrono::seconds(1));

  // Crash the program if the number is greater than 75
  if (randomNumber > 75) {
    std::cout << "Number is greater than 75. Crashing the program..."
              << std::endl;
    int *ptr = nullptr;
    *ptr = 42; // Dereferencing a null pointer to cause a crash
  }

  std::cout << "Program completed successfully." << std::endl;
  return 0;
}

File: exercises/ex2_occasional/occasional.f90
program occasional
  implicit none
  integer :: randomNumber
  integer, allocatable :: arr(:)
  real :: r

  call random_number(r)
  randomNumber = int(r * 100)

  print *, "Generated number:", randomNumber

  call decoyFunction()

  ! Sleep for 1 second
  call sleep(1)

  ! Crash the program if the number is greater than 75
  if (randomNumber > 75) then
    print *, "Number is greater than 75. Crashing the program..."
    deallocate(arr) ! attempt to deallocate unallocated array
  end if

  print *, "Program completed successfully."

  contains

  subroutine decoyFunction()
    ! This function does nothing but serves as a decoy
    implicit none
  end subroutine decoyFunction
end program occasional
